File: 7662.py
import heapq
import sys


def up_down_heap(operations):
    max_heap = []
    min_heap = []
    for operation in operations:
        operand, command = operation.split()
        if operand == "I":
            # insert
            heapq.heappush(max_heap, -int(command))
            heapq.heappush(min_heap, int(command))
        else:
            # deletion
            if min_heap:
                if command == '-1':
                    # min heap pop
                    min = heapq.heappop(min_heap)
                    max_heap.remove(-min)
                    heapq.heapify(max_heap)
                else:
                    max = heapq.heappop(max_heap)
                    min_heap.remove(-max)
                    heapq.heapify(min_heap)
    if not min_heap:
        return [0, 0]
    else:
        return [-max_heap[0], min_heap[0]]


T = int(input())

for _ in range(T):
    N = int(input())
    operations = []
    for op in range(N):
        operations.append(sys.stdin.readline().rstrip('\n'))

    max_h, min_h = up_down_heap(operations)
    if max_h == 0 and min_h == 0:
        print("EMPTY")
    else:
        print(max_h, min_h)

File: 7662.cpp
#include <algorithm>
#include <deque>
#include <iostream>
#include <queue>
#include <string>
#include <vector>

using namespace std;

pair<long long, long long> up_down_heap(
    vector<pair<char, long long>> operations) {
    priority_queue<long long, vector<long long>, greater<long long>> min_heap;
    priority_queue<long long, vector<long long>, greater<long long>>
        min_heap_dump;
    priority_queue<long long, vector<long long>, less<long long>> max_heap;
    priority_queue<long long, vector<long long>, less<long long>> max_heap_dump;

    for (pair<char, long long> operation : operations) {
        if (operation.first == 'I') {
            // insertion
            min_heap.push(operation.second);
            max_heap.push(operation.second);
        } else {
            // deletion
            if (min_heap.size() > 0) {
                if (operation.second == 1) {
                    // max_heap pop
                    min_heap_dump.push(max_heap.top());
                    max_heap.pop();

                } else {
                    // min_heap pop
                    max_heap_dump.push(min_heap.top());
                    min_heap.pop();
                }
            }
        }
        while (min_heap.size() > 0 && min_heap_dump.size() > 0 &&
               min_heap_dump.top() == min_heap.top()) {
            min_heap.pop();
            min_heap_dump.pop();
        }
        while (max_heap.size() > 0 && max_heap_dump.size() > 0 &&
               max_heap_dump.top() == max_heap.top()) {
            max_heap.pop();
            max_heap_dump.pop();
        }
    }
    if (min_heap.size() > 0) {
        return pair(max_heap.top(), min_heap.top());
    } else {
        return pair(0, 0);
    }
}

int main(void) {
    int T, N;
    long long num;
    char operand;
    scanf("%d", &T);

    while (T--) {
        scanf("%d", &N);
        vector<pair<char, long long>> operation;
        while (N--) {
            cin >> operand >> num;
            operation.push_back(pair(operand, num));
        }
        pair<long long, long long> answer = up_down_heap(operation);
        if (answer.first == 0 && answer.second == 0) {
            printf("EMPTY\n");
        } else {
            printf("%lld %lld\n", answer.first, answer.second);
        }
    }
}

/*
1
7
I 16
I -5643
D -1
D 1
D 1
I 123
D -1

1
9
I -45
I 653
D 1
I -642
I 45
I 97
D 1
D -1
I 333


2
7
I 16
I -5643
D -1
D 1
D 1
I 123
D -1
9
I -45
I 653
D 1
I -642
I 45
I 97
D 1
D -1
I 333
333 45
*/